### PR TITLE
Optimize Attestation Inclusion Calculations 

### PR DESCRIPTION
**Pull request by terenc3t**
Optimize attestation inclusion by doing the following:
1.) save `attestor` -> `attestation_slot_included` to a map during attestation processing
2.) use the map for inclusion slot rewards calculation

This optimization saves us from doing the following loop just to get attestor's inclusion slot (credit to @rauljordan)
```
for i in range(0, num_of_attestors):
    for j in range(0, num_of_attestors):
          participants = attestationParticipants() # expensive operation
          for z in range(0, len(participants)):
                # do something
```

### CODE REVIEW

File: beacon-chain/core/blocks/block_operations.go [hunk 1]
```diff
@@ -412,6 +427,36 @@ func ProcessBlockAttestations(
 		if err := VerifyAttestation(beaconState, attestation, verifySignatures); err != nil {
 			return nil, fmt.Errorf("could not verify attestation at index %d in block: %v", idx, err)
 		}
+
+		var IndicesInCommittee []uint64
+		// get the validator indices from the attestation using committees info cache.
```
Comment (terenc3t) [inline beacon-chain/core/blocks/block_operations.go]: Going to refactor this out into its own function

File: beacon-chain/core/balances/rewards_penalties_test.go [hunk 2]
```diff
@@ -471,17 +478,27 @@ func TestAttestationInclusionRewards_AccurateRewards(t *testing.T) {
...
+			LatestAttestations: pendingAtts,
+			LatestCrosslinks:   []*pb.Crosslink{{}},
 		}
-		state, err := AttestationInclusion(
+
+		_, err := blocks.ProcessBlockAttestations(state, &pb.BeaconBlock{
+			Body: &pb.BeaconBlockBody{
+				Attestations: atts,
+			},
+		}, false)
```
Comment (rauljordan) [inline beacon-chain/core/balances/rewards_penalties_test.go]: Comment on what this false is

File: beacon-chain/core/helpers/rewards_penalties.go [hunk 3]
```diff
@@ -17,6 +19,7 @@ import (
 //     """
 //     return min(state.validator_balances[idx], MAX_DEPOSIT)
 func EffectiveBalance(state *pb.BeaconState, idx uint64) uint64 {
+	fmt.Println(idx)
```
Comment (rauljordan) [inline beacon-chain/core/helpers/rewards_penalties.go]: Remove

File: beacon-chain/core/blocks/block_operations.go [hunk 4]
```diff
@@ -701,3 +751,25 @@ func verifyExit(beaconState *pb.BeaconState, exit *pb.VoluntaryExit, verifySigna
...
+		return slot, nil
+	}
+	return 0, fmt.Errorf("no inclusion slot for attestor %d", index)
+}
+
+// AttsInclusionDistance returns diff in slot of when an attestator's attestation gets submitted
+// and included.
+func AttsInclusionDistance(index uint64) (uint64, error) {
+	attsInclStore.RLock()
+	attsInclStore.RUnlock()
```
Comment (rauljordan) [inline beacon-chain/core/blocks/block_operations.go]: I think that these should be different mutexes

File: beacon-chain/core/balances/rewards_penalties_test.go [hunk 5]
```diff
@@ -4,6 +4,8 @@ import (
 	"reflect"
 	"testing"
 
+	"github.com/prysmaticlabs/prysm/beacon-chain/core/blocks"
```
Comment (rauljordan) [inline beacon-chain/core/balances/rewards_penalties_test.go]: spacing

**Pull request merged**